File: data-structures/trees/tries/Go/Trie.go
package main

import (
	"fmt"
)

// Node represents a node in the trie
type Node struct {
	children [26]*Node
	isEnd    bool
}

// Trie represents a trie and has a pointer to the root node
type Trie struct {
	root *Node
}

// Insert will take in a word and add it to the trie
func (t *Trie) Insert(word string) {
	wordLength := len(word)
	currentNode := t.root
	for i := 0; i < wordLength; i++ {
		charIndex := word[i] - 'a'
		if currentNode.children[charIndex] == nil {
			currentNode.children[charIndex] = &Node{}
		}
		currentNode = currentNode.children[charIndex]
	}
	currentNode.isEnd = true
}

// Search will take in a word and return true is that word is included in the trie
func (t *Trie) Search(word string) bool {
	wordLength := len(word)
	currentNode := t.root
	for i := 0; i < wordLength; i++ {
		charIndex := word[i] - 'a'
		if currentNode.children[charIndex] == nil {
			return false
		}
		currentNode = currentNode.children[charIndex]
	}
	if currentNode.isEnd {
		return true
	}
	return false
}

// Print displays all the words contained in the trie
func (t *Trie) Print() {
	t.root.print("", 0, false)
}

func (n *Node) print(word string, level int, sameLevel bool) {
	if n.isEnd {
		fmt.Println(word)
	}
	sameLevel = false
	for i := 0; i < 26; i++ {
		if n.children[i] != nil {
			if sameLevel && len(word) >= 1 {
				word = word[:len(word)-1]
			}
			sameLevel = true
			word += string(rune(i + 'a'))
			n.children[i].print(word, level+1, sameLevel)
		}
	}
}

// InitTrie will create a new trie
func InitTrie() *Trie {
	// Create a new trie with a root node
	return &Trie{root: &Node{}}
}

func main() {
	trie := InitTrie()

	words := []string{
		"car",
		"card",
		"cards",
		"cot",
		"cots",
		"trie",
		"tried",
		"tries",
		"try",
	}

	for _, word := range words {
		trie.Insert(word)
	}
	// trie.Print()

	for _, word := range words {
		if !trie.Search(word) {
			panic("this word should exists in the trie")
		}
	}

	fmt.Println(trie.Search("golang")) // false
	fmt.Println(trie.Search("carry"))  // false
}

File: data-structures/trees/tries/Java/Trie.java
class Trie {
    private Node root;
    
    public Trie() {
        root = new Node();
    }
    
    void insert(String word) {
        int wordLength = word.length();
        Node currentNode = root;
        for (int i = 0; i < wordLength; i++) {
            int charIndex = word.charAt(i) - 'a';
            if (currentNode.children[charIndex] == null) {
                currentNode.children[charIndex] = new Node();
            }
            currentNode = currentNode.children[charIndex];
        }
        currentNode.isEnd = true;
    }

    boolean search(String word) {
        int wordLength = word.length();
        Node currentNode = root;
        for (int i = 0; i < wordLength; i++) {
            int charIndex = word.charAt(i) - 'a';
            if (currentNode.children[charIndex] == null) {
                return false;
            }
            currentNode = currentNode.children[charIndex];
        }
        return currentNode.isEnd;
    }
    
    void print() {
        root.print("", 0, false);
    }
    
    class Node {
        private Node[] children;
        private boolean isEnd;
        
        public Node() {
            children = new Node[26];
            isEnd = false;
        }
        
        void print(String word, int level, boolean sameLevel) {
            if (isEnd) {
                System.out.println(word);
            }
            sameLevel = false;
            for (int i = 0; i < 26; i++) {
                if (children[i] != null) {
                    if (sameLevel && word.length() >= 1) {
                        word = word.substring(0, word.length()-1);
                    }
                    sameLevel = true;
                    word = word + (char)(i + 'a');
                    children[i].print(word, level+1, sameLevel);
                }
            }
        }
    }
    
    public static void main(String[ ] args) {
        Trie trie = new Trie();
        String[] words = {
            "car",
            "card",
            "cards",
            "cot",
            "cots",
            "trie",
            "tried",
            "tries",
            "try",
        };
        for (String word : words) {
            trie.insert(word);
        }
        // trie.print();

        for (String word : words) {
            if (!trie.search(word)) {
                System.out.println(word + " > this word should exists in the trie");
            }
        }

        System.out.println(trie.search("java")); // false
        System.out.println(trie.search("carry")); // false
    }
}